Initial integration of nimls

yyoncho opened the pull request:
- Added support for the experimental `nimls` language server. The server will be
developed in the following repo https://github.com/nim-lang/langserver . The
server design follows https://github.com/nim-lang/langserver/blob/master/docs/design/LangServArchitecture.md

Check https://github.com/nim-lang/langserver/pull/3

yyoncho committed:
Initial integration of nimlangserver

- Added support for the experimental `nimlangserver` language server. The server will be
developed in the following repo https://github.com/nim-lang/langserver . The
server design follows https://github.com/nim-lang/langserver/blob/master/docs/design/LangServArchitecture.md

saem commented on .vscodeignore:
```diff
@@ -11,5 +11,4 @@ src/**
 .gitignore
 tsconfig.json
-tslint.json
-node_modules/**
```
Why remove this and are the line endings incorrect?

yyoncho replied:
the vsix has to package `"vscode-languageclient": "7.0.0"`

saem replied:
It's been a while since I looked at vscodeignore, that'll preclude is from including the client?

yyoncho replied:
Yes, `vsce` looks into this file and based on it includes or does not a particular file/directory.

saem replied:
Got it, can we restrict it so it only looks at that module, node modules is huge

saem commented on README.md:
```diff
@@ -104,6 +103,46 @@ The following Visual Studio Code settings are available for the Nim extension.
+
+Similar to `vscode-nim` [Nim Language Server](https://github.com/Nim-lang/langserver) uses `nimsuggest` under the
+hood and to control the way `nimsuggest` is created you can use
+`nim.rootConfig`. Here it is a sample config:
```
```suggestion
`nim.rootConfig`. Here is a sample config:
```

saem commented on README.md:
````diff
@@ -104,6 +103,46 @@ The following Visual Studio Code settings are available for the Nim extension.
+```
+
+Similar to `vscode-nim` [Nim Language Server](https://github.com/Nim-lang/langserver) uses `nimsuggest` under the
+hood and to control the way `nimsuggest` is created you can use
````
Why isn't this using the existing project and project mapping config?

yyoncho replied:
TBO I was not aware of this config. I will check both can be adjusted to use the same set of settings.

saem replied:
Nailed it

saem commented on README.md:
```diff
@@ -104,6 +103,46 @@ The following Visual Studio Code settings are available for the Nim extension.
+
+* (Re)starting `nimsuggest` subprocess
+
+Use `Command palette...`  -> `Source actions` -> `Restart nimsuggest`
```
This should be under commands and the restart should work for either mode.

yyoncho replied:
This is a standard language server protocol approach and it will work fine across all clients without the need to create custom commands in all clients.

saem replied:
I meant the headings commands in the readme.

Also, that it should look at the config, if there is epc nimsuggest it should restart those or of it's nimls it should restart it

yyoncho replied:
> Also, that it should look at the config, if there is epc nimsuggest it should restart those or of it's nimls it should restart it

The command comes as a result of `textDocument/codeActions` and it is implemented on the server thus it does not have access to the client state. I cannot think of an easy way to make restart work for both.

saem replied:
https://code.visualstudio.com/api/extension-guides/command#registering-a-command

You can register a listener, it'll listen to the the command, just guard it on the config

saem commented on src/nimvscode.nim:
```diff
@@ -32,11 +33,14 @@ from nimSuggestExec import extensionContext, initNimSuggest,
 var fileWatcher {.threadvar.}: VscodeFileSystemWatcher
 var terminal {.threadvar.}: VscodeTerminal
+var client {.threadvar.}: VscodeLanguageClient
+var installPerformed {.threadvar.}: bool
```
This should be in the extension state

yyoncho replied:
sure

saem commented on src/nimvscode.nim:
```diff
@@ -313,7 +357,9 @@ proc activate*(ctx: VscodeExtensionContext): void =
-  if config.getBool("enableNimsuggest"):
+  if config.getBool("enableLsp"):
+    startLanguageServer(true)
+  elif config.getBool("enableNimsuggest"):
```
We shouldn't have two separate flags, just one property which is an enum:
- nimsuggest
- nimls
- none

yyoncho replied:
won't that be a breaking change?

saem replied:
I'm trying to recall where I saw it, but there is a way to handle migration of config in vscode such that it's not. 🤔

yyoncho replied:
I can make it like that:

have a property `backend` which can be the 3 values listed above. The default is nimsuggest and the when `backend == nimsuggest && enableNimsuggest == false` assume that backend is `none`

yyoncho replied:
Alternatively, we may keep both flags given the fact that most likely direct integration with nimsuggest will be deprecated in favour of lsp.

saem replied:
The value property works, let's not call it backend as that is already a property. Otherwise sounds good.

saem commented on src/nimvscode.nim:
```diff
@@ -292,6 +296,46 @@ proc clearCachesCmd(): void =
   let config = vscode.workspace.getConfiguration("files")
   discard clearCaches(config.getStrBoolMap("watcherExclude", defaultIndexExcludeGlobs))
 
+proc initializeClient(nimls: cstring) =
```
Please put these in their own module, then export/import the minimum

yyoncho replied:
sure.

saem reviewed (changes requested):
This is a good start, few changes and it should fit in nicely.

yyoncho commented:
@saem 

> Got it, can we restrict it so it only looks at that module, node modules is huge

`vsce` does that by default. It does not pick dev deps.

> You can register a listener, it'll listen to the the command, just guard it on the config

What do you mean by guarding it?

saem commented:
> > You can register a listener, it'll listen to the the command, just guard it on the config
> 
> What do you mean by guarding it?

If check to ensure the backend is epc vs not.

yyoncho commented:
> > > You can register a listener, it'll listen to the the command, just guard it on the config
> > 
> > 
> > What do you mean by guarding it?
> 
> If check to ensure the backend is epc vs not.

There is no need for that, this guarding is performed by VScode itself. The epc won't register code action provider at all.

saem commented:
My point is that the behavior should be uniform. So one can restart nimsuggest in either case (epc or nimls), that way there isn't a bifurcated experience.

yyoncho commented:
IMHO having a bifurcated experience is inevitable given the fact that more and more functionality that won't be present in bare nimsuggest integration will be added to nimls. I can investigate how to make the command work for both - my concern is that this will be a wasted effort.

saem commented on README.md:
```diff
@@ -87,10 +87,11 @@ The following Visual Studio Code settings are available for the Nim extension.
 * `nim.lintOnSave` - perform the project check for errors on save
 * `nim.project` - optional array of projects file, if nim.project is not defined then all nim files will be used as separate project
 * `nim.licenseString` - optional license text that will be inserted on nim file creation
+* `nim.provider` - specifies the backend to use for language features. It can be
```
Awesome, thanks

saem commented on src/nimLsp.nim:
```diff
@@ -0,0 +1,45 @@
+      vscode.window.showInformationMessage("Unable to find nimlangserver, trying to install it via 'nimble install nimlangserver --accept'")
+      state.installPerformed = true
+      discard cp.exec(
+        getNimbleExecPath() & " install nimlangserver --accept",
```
On my phone so I can't write the code out as a suggestion could you:

Assign this command build to a let in an earlier line, then use that in the show information message string so they don't go out of sync.

yyoncho replied:
sure

saem commented on src/nimLsp.nim:
```diff
@@ -0,0 +1,45 @@
+        getNimbleExecPath() & " install nimlangserver --accept",
+        ExecOptions{},
+        proc(err: ExecError, stdout: cstring, stderr: cstring): void =
+          console.log("Nimble install finished, checking if nimlangserver is already present.")
```
```suggestion
          console.log("Nimble install finished, validating by checking if nimlangserver is present.")
```

saem commented on src/platform/languageClientApi.nim:
```diff
@@ -0,0 +1,34 @@
+import js/[jsPromise, asyncjs, jsre, jsNode]
+
+export jsffi, jsPromise, jsNode
+
```
Could you add a comment with a link to the API docs, that way others can quickly get to the right place if they need to confirm, fix, update, whatever this shim.

yyoncho replied:
ok

saem commented on src/tools/nimBinTools.nim:
```diff
@@ -25,13 +25,17 @@ proc getBinPath*(tool: cstring): cstring =
 
-    binPathsCache[tool] = pathParts.mapIt(
+    let paths = pathParts.mapIt(
         block:
```
Ugh, I forgot about gems like these, sorry you had to work with that ugly bit.

saem reviewed (changes requested):
Last mile at this point.

Is listening to the restart command clearer?

yyoncho commented:
Force pushed.

yyoncho commented:
ping

saem commented:
Had read through, last bit is register a restart command for nimsuggest so that it's uniform.

yyoncho commented:
Ok, I will try to implement it, although it looks like it will do more harm than good.

yyoncho committed:
Addressing the comments

yyoncho commented:
Added restart command for `nimsuggest`. Also, I changed a bit the wording in the readme to make it a bit more evident that one should expect a different set of functionality from `lsp` and `nimsuggest` modes in addition to the difference in how both behave.

saem commented:
Sorry, for some reason I didn't see the notification for GitHub. Awesome, thanks.

yyoncho commented:
> Sorry, for some reason I didn't see the notification for GitHub. Awesome, thanks.

Thank you! Can you do a release?

yyoncho commented:
> > Sorry, for some reason I didn't see the notification for GitHub. Awesome, thanks.
> 
> Thank you! Can you do a release?

Any update?

saem commented:
https://github.com/saem/vscode-nim/commit/17070382e277188cf0a64f9fdcee3b6d28bdc501
